proposals/exception-handling/Exceptions.md

```diff
@@ -205,99 +193,166 @@ flushed, the embedder defines how to handle uncaught exceptions. Otherwise, the
 
 If control is transferred to the body of a catch block, and the last instruction
 in the body is executed, control then exits the try block.
 
-If the selected catch block does not throw an exception, it must yield the
-value(s) expected by the corresponding catching try block. This includes popping
-the caught exception.
+In case of a try-unwind block, the control is transfered to the body of the
+`unwind` block. After executing instructions within the `unwind` block, the
+exception is automatically rethrown when the control reaches the `end`
+instruction. As in the case of the `catch_all` block, the exception arguments
+are not copied onto the operand stack.
+
+If the selected catch/unwind block does not throw an exception, it must yield
+the value(s) expected by the corresponding catching try block. This includes
+popping the caught exception.
 
 Note that a caught exception can be rethrown using the `rethrow` instruction.
 
 ### Rethrowing an exception
 
-The `rethrow` instruction takes the exception associated with the `exnref` on
-top of the stack, and rethrows the exception. A rethrow has the same effect as a
-throw, other than an exception is not created. Rather, the referenced exception
-on top of the stack is popped and then thrown. The `rethrow` instruction traps
-if the value on the top of the stack is null.
+The `rethrow` instruction can only appear in the body of a catch/catch_all block
+but not within the body of an unwind block. It always re-throws the exception
+caught by an enclosing catch block.
 
-### Exception data extraction
+Associated with the `rethrow` instruction is a _label_. The label is used to
+disambiguate which exception is to be rethrown, when inside nested catch blocks.
+The label is the relative block depth to the corresponding try block for which
+the catching block appears.
 
-The `br_on_exn` instruction is a conditional branch that checks the exception
-tag of an exception on top of the stack, in the form of:
+For example consider the following:
 
 ```
-br_on_exn label except_index
+try $l1
+  ...
+catch 1
+  ...
+  block
+    ...
+    try $l2
+      ...
+    catch 2
+      ...
+      try $l3
+        ...
+      catch 3
+        ...
+        rethrow N (or label)
+      end
+    end
+  end
+  ...
+end
+```
+
+In this example, `N` is used to disambiguate which caught exception is being
+rethrown. It could rethrow any of the three caught expceptions. Hence,
+`rethrow 0` corresponds to the exception caught by `catch 3`, `rethrow 1`
+corresponds to the exception caught by `catch 2`, and `rethrow 3` corresponds
+to the exception caught by `catch 1`. In wat format, the argument for the
+`rethrow` instructions can also be written as a label, like branches. So
+`rethrow 0` in the example above can also be written as `rethrow $l3`.
+
+Note that `rethrow 2` is not allowed because it does not reference a `try`
+instruction. Rather, it references a `block` instruction, so it will result in a
```

Comment:
```suggestion
Note that `rethrow 2` is not allowed because it does not reference a catch
block. Rather, it references a `block` instruction, so it will result in a
```